### PR TITLE
Glossaries

### PR DESCRIPTION
**Pull request by haggaie**
The pull-request fixes a glossaries warning, and adds an example text using glossaries. It also checks the behavior with a long word.

### CODE REVIEW

File: misc/iitthesis.cls [hunk 1]
```diff
@@ -607,6 +607,7 @@
   \renewcommand{\glossarysection}[2][]{}
   \newcommand{\makenotationAndAbbreviations}{%
    \chapter*{Notation and Abbreviations}
+   \label{chap:abbrev}
```
Comment (eyalroz) [inline misc/iitthesis.cls:610]: Why not `\label{chap:notations-and-abbrevations}` ?

File: main/prelims.tex [hunk 2]
```diff
@@ -11,3 +11,25 @@ \section{Some section}
...
+  letters), the \verb|\gls{tla}| macro of the \texttt{glossaries-extra} package
+  generates its full form, with its acronym in parentheses. The acronym is defined
+  in the \texttt{abbrevs.tex} file. The next time a \gls{tla} is used, only its
+  short form is printed. \Glspl{etla} can be capitlized or use the plural form
+  with macros such as \verb|\Gls| or \verb|\glspl|.
+
+  In addition, the \verb|\gls| macro adds the pages where a term is used to the
+  \nameref{chap:abbrev} \autoref{chap:abbrev}.
+\else
+  The Notation and Abbreviations Chapter is disabled.
```
Comment (eyalroz) [inline main/prelims.tex:34]: I understand the motivation to have this conditional code. But - this is a sample thesis. Let's assume a relatively naive author and just take the `\ifabbreviations` path here and drop the the `\else` case.

File: main/prelims.tex [hunk 3]
```diff
@@ -11,3 +11,25 @@ \section{Some section}
 The \emph{von Neumann model} of a computer, also known as the \emph{Princeton architecture} is an architecture for digital computers, which consists of a processing units, containing an ALU and processing registers; a control unit consisting of an instruction register and a program counter; a memory unit which stores both data and instructions; and input-and-output mechanisms.
 \end{definition}
 
+\section{Acronyms and abbreviations}
+
+\newif\ifabbreviations
```
Comment (eyalroz) [inline main/prelims.tex:16]: We should assume authors know very little LaTeX, and not suggest they rely on conditional compilation and use of LaTeX booleans...

**Review by eyalroz — CHANGES_REQUESTED** [review]
Thanks :-)

Please have a look at my individual comments.

File: misc/iitthesis.cls (hunk 1 above)
Comment (haggaie) [reply misc/iitthesis.cls:610]: Sure, I'll change that.

File: main/prelims.tex (hunk 2 above)
Comment (haggaie) [reply main/prelims.tex:34]: The problem is that the \gls macro is not defined when `noabbrevs` is used.

Comment (eyalroz) [reply main/prelims.tex:34]: It's not a problem. It's ok that the `\gls`'s will trigger errors if `noabbrevs` is added. In fact, I would mention it in the text, i.e. say that in writing this paragraph we are using the `\gls` command, which is not necessarily available if `noabbrevs` is specified.

**Comment by eyalroz** [comment]
I'm merging into the branch, but we're not there yet:

1. You're silently requiring the `longtable` package; package requirements should be explicit. I could fix that, but...
2. How sure are we that `longtable` doesn't clash with other packages? Here's a niche example: [a clash with `caption` in some cases](https://tex.stackexchange.com/q/245980/5640). There was also a clash with `bigfoot` around 2015 which was resolved. How sure are you this is a safe choice?
3. I don't think 2cm is sufficient for the defined term. First, I'm not even certain the width should be fixed, but if it is fixed - let's make it wider. I'd say at least 3cm. I don't suppose we could have some sort of dynamic computation of widths here, right?
4. There's not enough space between the term column and the next column.
5. I don't like the colons. These are not sentences. Why not just drop them?
6. If a term exceeds the column length, we get a rather weird layout: Its first line will be indented, and we'll have almost no space between the hypens and the colon-column. The space issue goes away if we remove the colons, but the indentation remains. Try uncommenting the `\keepUnusedAbbreviations` command and you'll see this in "action"...
7. Multi-line definitions have no indentations, i.e. we don't have good visual separation between the last line of one definition and the first line of the next.

I realize some of this is a bit nitpicky (e.g. item 7), but if we're already doing this, let's try to round it out.

**Comment by eyalroz** [comment]
So, this was rebase-merged, even though it looks like it wasn't do to some other commit changes.

**Comment by eyalroz** [comment]
Two more issues:

1. If you refer to a term on multiple pages, they will _all_ appear on the list page. That's not acceptable... it makes it effectively a partial Index, which isn't what we want. We need to have only the first appearance be listed (if at all).
2. There's this command: `\setabbreviationstyle[acronym]{long-short}` in `abbrevs.tex`. I don't like it there. Either we make this the fixed default, bringing it into the document class; drop it altogether; or present several alternatives.

**Comment by haggaie** [comment]
> I'm merging into the branch, but we're not there yet:
> 
> 1. You're silently requiring the `longtable` package; package requirements should be explicit. I could fix that, but...
> 2. How sure are we that `longtable` doesn't clash with other packages? Here's a niche example: [a clash with `caption` in some cases](https://tex.stackexchange.com/q/245980/5640). There was also a clash with `bigfoot` around 2015 which was resolved. How sure are you this is a safe choice?
> 3. I don't think 2cm is sufficient for the defined term. First, I'm not even certain the width should be fixed, but if it is fixed - let's make it wider. I'd say at least 3cm. I don't suppose we could have some sort of dynamic computation of widths here, right?
> 4. There's not enough space between the term column and the next column.
> 5. I don't like the colons. These are not sentences. Why not just drop them?
> 6. If a term exceeds the column length, we get a rather weird layout: Its first line will be indented, and we'll have almost no space between the hypens and the colon-column. The space issue goes away if we remove the colons, but the indentation remains. Try uncommenting the `\keepUnusedAbbreviations` command and you'll see this in "action"...
> 7. Multi-line definitions have no indentations, i.e. we don't have good visual separation between the last line of one definition and the first line of the next.
> 
> I realize some of this is a bit nitpicky (e.g. item 7), but if we're already doing this, let's try to round it out.

We don't have to use a table style. The reasons I did that were (1) trying to match the previous format and (2) copying an example of how to put a description and an acronym on the same line. QED acronym had both, so that was also part of trying to keep up with the existing example.

In most cases, I think just having the long version of the acronym is enough, so perhaps we can use one of the standard styles that come with glossaries/glossaries-extra, and avoid longtable and other style issues.

**Comment by haggaie** [comment]
> Two more issues:
> 
> 1. If you refer to a term on multiple pages, they will _all_ appear on the list page. That's not acceptable... it makes it effectively a partial Index, which isn't what we want. We need to have only the first appearance be listed (if at all).

Sure, the `indexonlyfirst` option should do that.

> 2. There's this command: `\setabbreviationstyle[acronym]{long-short}` in `abbrevs.tex`. I don't like it there. Either we make this the fixed default, bringing it into the document class; drop it altogether; or present several alternatives.

I think this can be the default, and users can override it by using this command in their code if they want (before the `\newacronym` commands).

**Comment by eyalroz** [comment]
> Sure, the `indexonlyfirst` option should do that.

Yeah, but... then all glossaries (including any which the thesis author decides to add) will have this option. I'll do it, but I wish I could apply it at the single-glossary level.

> > 2. There's this command: `\setabbreviationstyle[acronym]{long-short}` in `abbrevs.tex`. I don't like it there. Either we make this the fixed default, bringing it into the document class; drop it altogether; or present several alternatives.
> 
> I think this can be the default, and users can override it by using this command in their code if they want (before the `\newacronym` commands).

OK.

**Comment by eyalroz** [comment]
> In most cases, I think just having the long version of the acronym is enough, so perhaps we can use one of the standard styles that come with glossaries/glossaries-extra, and avoid longtable and other style issues.

Well, that's a bit drastic... but if you want suggest something concrete, then let's consider that. The `longtable` arrangement is not perfect, but it's still much better than what the class had before.

**Comment by haggaie** [comment]
> Yeah, but... then all glossaries (including any which the thesis author decides to add) will have this option. I'll do it, but I wish I could apply it at the single-glossary level.

Looking at the manual, there's a way to configure indexing first use only per "category", and also to turn indexing on or off per call to `\gls`. They recommend the latter method as sometimes the first use isn't where you want the index to point. You could use

    \GlsXtrSetDefaultGlsOpts{noindex}

to make the default without index, and then use `\gls[noindex=false]{...}` to index a specific entry.

**Comment by haggaie** [comment]
> > In most cases, I think just having the long version of the acronym is enough, so perhaps we can use one of the standard styles that come with glossaries/glossaries-extra, and avoid longtable and other style issues.
> 
> Well, that's a bit drastic... but if you want suggest something concrete, then let's consider that. The `longtable` arrangement is not perfect, but it's still much better than what the class had before.

Here's how the default style looks like. I'm not sure, but it reminds me of a description environment:

image

The problem with the default is that it only shows one field for the definition, either the long form of the acronym, or a description for entries, but not both.

**Comment by eyalroz** [comment]
> Looking at the manual, there's a way to configure indexing first use only per "category"

Well, we don't need per-category changes, we need per-glossary changes.

In fact, I had a look in some of the (many) manuals myself. It turns out that, so far, we've been using the main glossary. Which is something we should actually be avoiding, to not interfere with the document author's possible use of it.

So, I've made a few extra changes to the abbrevs-2022-01 branch, so that we now use 

> and also to turn indexing on or off per call to `\gls`.

But we want all `\gls` calls to be the same, in case they get reordered.

> 
> ```
> \GlsXtrSetDefaultGlsOpts{noindex}
> ```
> 
> to make the default without index, and then use `\gls[noindex=false]{...}` to index a specific entry.

We can't do that. The defaults for the package need to stay without change. Everything we do needs to apply just to our own glossary.

Maybe I'll ask on tex.stackexchange.com about how to achieve the first-appearance-only page listing; but I'll need a minimal testcase for that.

**Comment by eyalroz** [comment]
@haggaie I just asked [this question](https://tex.stackexchange.com/questions/630541/how-can-i-get-the-effect-of-indexonlyfirst-for-a-single-glossary).

**Comment by haggaie** [comment]
Great. To be honest, I'm a little confused about the distinction between a
glossary and a category. I'll have to read more to understand when you
would want to use each.

בתאריך יום ו׳, 28 בינו׳ 2022, 17:19, מאת Eyal Rozenberg ‏<
***@***.***>:

**Comment by eyalroz** [comment]
Glossaries and categories are orthogonal: A `\newglossaryentry` can be assigned a type and a category (perhaps multiple categories?)

A glossary will show all entries whose type is the glossary' type.

**Comment by haggaie** [comment]
I see. And your goal is to have one glossary from the thesis class, which must be the non-default type, and another glossary created by the author, which would use the `general` type?

Would using a non-default category (or set of categories, if you want to have, e.g., acronyms and abbreviations as different categories) solve it, as schtandard suggested?

**Comment by eyalroz** [comment]
My hope is for a solution which would not hamper or impede any use by the thesis author of the default and any other glossaries, or an index. The thing is, that the thesis author may theoretically want the same term both in another glossary and in notations-and-abbreviations - and the category-based styling doesn't solve that. 

Still, maybe that's good enough for now, as, TBH, it's probably rare for Technion PhD theses to have indices. Maybe I'll lazy-out and accept this solution.

**Comment by eyalroz** [comment]
Please have a look at #46 and let me know what you think. It's not an ideal state of affairs, but it's sorta-kinda passable.

**Pull request closed without merging**